### PR TITLE
impl #397, new composition dialog on facebook

### LINKED ISSUE #397: [Demand] Post Composition UI - Temporary Consolidation
## Metadata

Field | Value
--- | ---
Previous Stage | ...
Figma Link | [figma.com/...](https://www.figma.com/file/nDyLQp036eHgcgUXeFmNA1/Post-Composition-v1?node-id=12%3A183)
Expected UX prototype due | 2019-11-18
Actual UX prototype release | 2019-11-18
Expected engineering due | 2019-11-22

## Abstract

Before the real consolidation of post composition UI may be introduced, we need a framework to describe UI behaviors. This is especially useful since multiple-network support is coming into reality.

### Definitions of Classes

The composing state ephemerality may fall in 3 classes:

Class | Flag 1: Manual Invocation Required & Must Exit Composing State Before Doing Whatever Else | Flag 2: Automatically Exiting On Textarea Blur Event
--- | --- | ---
Static | False | N/A
Hybrid | True | False
Ephemeral | True | True

Notes:

- Pseudo-content mock boxes are considered Hybrid.
- Generally, no network should choose Ephemeral. But this case should be included in the list.

### Known Behavior

Network | Class
--- | ---
Facebook (Desktop Web) | Hybrid
Twitter (Desktop Web) | Static + Hybrid

### Behavior Consistency (Deprecated)

The invocation and the dismissal of Maskbook post composition UI should align to the behavior of the composition UI at the place where the native post composition UI is displayed.

If the native post composition UI is Hybrid, Maskbook post composition UI should be dismissed as soon as the native post composition UI is dismissed. However, if the native post composition UI is Static, Maskbook post composition UI should be manually invoked upon focusing, and there should be no way to dismiss Maskbook post composition UI there. If such persistence becomes a problem, we will provide a way to dismiss Maskbook post composition UI for Static scenarios.

### Handling Hybrid Composing Areas

The invocation of Maskbook composing modal is automatic upon the invocation event of native composing box. When exiting Maskbook composing modal, the native composing area is automatically focused, in order to prevent cyclical invocation of Maskbook composing modal.

In other words, Maskbook composing modal will be automatically invoked again only if the user exit native composing state and re-enter native composing state.

### Handling Static Composing Areas

It will be a bit tricky to deal with Static native composing areas.

When a static composing area is focused by the user, we will check its `data-maskbook-composing-box-already-invoked` attribute. If the value is `false`, Maskbook composing modal will be prompted and the attribute will be set to be `true`; else, nothing will happen. So that one Static composing area can only automatically trigger prompting Maskbook composing modal once.

### Reentry to Maskbook Composing Modal

We will add network-specific elements for the user to manually open Maskbook composing modal. This is for both cases.

image

Styling should be place-specific. Network-specific styling is not enough.

## UI Elements

Now we have to display a stand-alone composing box besides the original one instead of displaying it over the original one.

### Figma Preview

image

### Composing Box

A generic `textarea` element.

### Recipients Selection Bar

Each network, according to its friendship model, has different default options.

These options should be displayed with [Filter Chips](https://material.io/components/chips/#filter-chips) according to Material Design specifications.

Facebook:

Place | Options
--- | ---
My News Feed | Global, All Friends
Friend Timeline | Global, All Friends, This Friend
Real Group | Global, Group Members
Brand Page Manager | Global, Followers

Twitter:

Place | Options
--- | ---
Timeline | Global, Followers, Following, Following Each Other
Reply | Global, Followers, Following, Following Each Other

When replying, OP and mentioned users are always automatically and implicitly added.

#### Notes

- Global, Group Members, and Following Each Other do not have to be implemented for now. They are only being documented for future references.

### Advanced Recipients Selection (Debug Mode Only)

image

One more option in recipients options.

The text should be `Specific Friends (${specificRecipients.length} selected)`. By clicking the option, another modal, Specific Recipients Selection Modal, is displayed over the current modal.

image

In the Specific Recipients Selection Modal, the user may toggle the checkboxes besides the friends. Click "Done" button to close the modal.

(But which Material Design component should we use?)

### Finish Button

A button showing "Finish".

## Matching Network-Specific Styles (Suspended)

// This part can be delayed to a future stage

Font-family, font-size, line-height, letter-spacing, and color should match the current network. The Finish button should also look like the Post/Tweet button.

Styling parameters should be customizable with a simple CSS configuration.

## Editing

For networks which support editing existing posts, we should provide that compatibility. IV and post key should remain unchanged, as long as IV is 128-bit long and is used as an equivalent to UUIDv4 for post-identifying purposes.

### PR DESCRIPTION
**Pull request by Jack-Works**
close #397 

image

image

image

### CODE REVIEW

File: src/components/InjectedComponents/PostDialog.tsx [hunk 1]
```diff
@@ -342,30 +345,25 @@ export function PostDialog(props: PostDialogProps) {
...
         setOpen(false)
-    }, [])
+    }, [setOpen])
     //#endregion
     //#region My Identity
     const identities = useMyIdentities()
-    const [open, setOpen] = useState(false)
     useEffect(() => {
-        const onChange = ({ reason, open }: CompositionEvent) => {
+        return MessageCenter.on('compositionUpdated', ({ reason, open }: CompositionEvent) => {
```
Comment (differui) [inline src/components/InjectedComponents/PostDialog.tsx]: Should we remove the event listener after component destroyed?

File: src/components/InjectedComponents/PostDialog.tsx [hunk 2]
```diff
@@ -342,30 +345,25 @@ export function PostDialog(props: PostDialogProps) {
             setPostBoxContent(makeTypedMessage(''))
             setCurrentShareTarget([])
             getActivatedUI().typedMessageMetadata.value = new Map()
-        }, []),
+        }, [setOpen]),
```
Comment (differui) [inline src/components/InjectedComponents/PostDialog.tsx]: Any problems does this code fixed?

Comment (SunriseFox) [reply src/components/InjectedComponents/PostDialog.tsx]: very likely these were added by eslint...

File: src/components/InjectedComponents/PostDialog.tsx (hunk 1 above)
Comment (Jack-Works) [reply src/components/InjectedComponents/PostDialog.tsx]: the `on` function returns a function that will remove the listener.

File: src/components/InjectedComponents/PostDialog.tsx (hunk 2 above)
Comment (Jack-Works) [reply src/components/InjectedComponents/PostDialog.tsx]: It because now `open, setOpen` can be provided from props and no longer to be a local hook.

**Comment by Jack-Works** [comment]
@guanbinrui a type error here

**Comment by guanbinrui** [comment]
> @guanbinrui a type error here

Fixed!

**Comment by neruthes** [comment]
Finally! 🎉

**Pull request merged**